exec/src/klio_exec/runners/pubsub_message_manager.py

```diff
@@ -0,0 +1,284 @@
+        #    was moved to the top of this module.
+        # 2. Import statements adjusted to import module and not objects
+        #    according to the google style guide.
+        # 3. The functionalty we needed to override, which skips auto-acking
+        #    consumed pubsub messages, and adds them to the MessageManager
+        #    to handle deadline extension and acking once done.
+
+        def _get_element(ack_id, message):
+            parsed_message = beam_pubsub.PubsubMessage._from_message(message)
+            if (
+                timestamp_attribute
+                and timestamp_attribute in parsed_message.attributes
+            ):
+                rfc3339_or_milli = parsed_message.attributes[
+                    timestamp_attribute
+                ]
+                try:
+                    timestamp = beam_timestamp.Timestamp(
+                        micros=int(rfc3339_or_milli) * 1000
+                    )
+                except ValueError:
+                    try:
+                        timestamp = beam_timestamp.Timestamp.from_rfc3339(
+                            rfc3339_or_milli
+                        )
+                    except ValueError as e:
+                        raise ValueError("Bad timestamp value: %s" % e)
+            else:
+                timestamp = beam_timestamp.Timestamp(
+                    message.publish_time.seconds,
+                    message.publish_time.nanos // 1000,
+                )
+
+            # TODO: either use klio.mssage.serializer.to_klio_message, or
+            # figure out how to handle when a parsed_message can't be parsed
+            # into a KlioMessage
+            kmsg = klio_pb2.KlioMessage()
+            kmsg.ParseFromString(parsed_message.data)
+            entity_id = kmsg.data.element.decode("utf-8")
+
+            pmsg = PubSubKlioMessage(ack_id, entity_id)
+            self.message_manager.add(pmsg)
+
+            return timestamp, parsed_message
+
+        try:
+            response = self.sub_client.pull(
+                self._sub_name, max_messages=1, return_immediately=True
+            )
+            results = [
+                _get_element(rm.ack_id, rm.message)
+                for rm in response.received_messages
+            ]
+
+        finally:
+            self.sub_client.api.transport.channel.close()
+
+        return results
+
+
+class KlioTransformEvaluatorRegistry(
+    transform_evaluator.TransformEvaluatorRegistry
+):
+    def __init__(self, *args, **kwargs):
+        super(KlioTransformEvaluatorRegistry, self).__init__(*args, **kwargs)
+        self._evaluators[
+            direct_runner._DirectReadFromPubSub
+        ] = KlioPubSubReadEvaluator
+
+
+class KlioAckInputMessage(beam.DoFn):
+    def process(self, element):
+        kmsg = klio_pb2.KlioMessage()
+        kmsg.ParseFromString(element)
+        entity_id = kmsg.data.element.decode("utf-8")
+
+        msg = ENTITY_ID_TO_ACK_ID.get(entity_id)
```

Comment:
I know you distinguish the Klio message as `kmsg` above but it did take a sec to realize `msg` is a `PubSubKlioMessage`. Totally non-blocking but it was helpful calling the variable `pmsg` like you did above since theres so many message references getting thrown around 😬